src/main/flight/pid.c

```diff
@@ -436,28 +439,94 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float calcHorizonLevelStrength(void)
+    }
+    pidRuntime.previousAngle[axis] = currentAngle;
+    //angle loop feedforward
+#ifdef USE_FEEDFORWARD
+    float angleFeedforwardInput = pidRuntime.angleFeedforward[axis];
+    if (newRcFrame){
+        const float feedforwardJitterFactor = pidRuntime.feedforwardJitterFactor;
+        const float rcCommandDelta = fabsf(getRcCommandDelta(axis));
+        uint8_t duplicateCount = getFeedforwardDuplicateCount(axis);
+        if (rcCommandDelta) {
+            const float rxRateHz = 1e6f / getCurrentRxRefreshRate();
+            angleFeedforwardInput = (angleTarget - pidRuntime.previousAngleTarget[axis]) * rxRateHz;
+            //duplicate handling
+            if (duplicateCount) {
+                    angleFeedforwardInput /= duplicateCount + 1;
+                }
+            //reduce feedforward if close to angle limit
+            if (fabsf(currentAngle) > 0.95f * levelAngleLimit && angleFeedforwardInput < 3.0f * pidRuntime.angleFeedforward[axis]) {
+                    // approaching max stick position so zero out feedforward to minimise overshoot
+                    angleFeedforwardInput = 0.0f;
+                }
+            //PT1 filter for a derivative term (Feedforward)
+            angleFeedforwardInput = pidRuntime.angleFeedforward[axis] + pidRuntime.feedforwardSmoothFactor * (angleFeedforwardInput - pidRuntime.angleFeedforward[axis]);
+            //jitter attenuation
+            float jitterAttenuator = 1.0f;
+            if (feedforwardJitterFactor) {
+                if (rcCommandDelta < feedforwardJitterFactor) {
+                    jitterAttenuator = MAX(1.0f - (rcCommandDelta / feedforwardJitterFactor), 0.0f);
+                    jitterAttenuator = 1.0f - jitterAttenuator * jitterAttenuator;
+                }
+            }
+            angleFeedforwardInput *= jitterAttenuator;
+            //angle feedforward averaging at RCrate inactive for low speed links and 2-point for fast links
+            if (pidRuntime.feedforwardAveraging){
+                pidRuntime.angleFeedforward[axis] = laggedMovingAverageUpdate(&pidRuntime.angleFeedforwardAvg[axis].filter, pidRuntime.angleFeedforward[axis]);
+            }
+            pidRuntime.previousAngleTarget[axis] = angleTarget;
+            //feedforward transition
+            angleFeedforwardInput *= pidRuntime.feedforwardTransitionFactor > 0 ? MIN(1.0f, getRcDeflectionAbs(axis) * pidRuntime.feedforwardTransitionFactor) : 1.0f;
+        } else {
+            if (duplicateCount > 1) {
+                angleFeedforwardInput = 0;
+            }
+        }
+    }
+    //angle feedforward filter at the PID loop rate
+    float angleFeedforward = pt3FilterApply(&pidRuntime.angleFeedforwardPt3[axis], angleFeedforwardInput);
+    pidRuntime.angleFeedforward[axis] = angleFeedforwardInput;
+    angleFeedforward *= pidRuntime.levelFeedforwardGain;
+    pidRuntime.angleDerivative[axis] = currentAngleDerivative;
+    currentAngleDerivative *= pidRuntime.angleDerivativeGain;
```

Comment:
can be deleted, done up the top at PID loop rate.
```suggestion
```

Reply:
agreed.